source/custom/nixnet_service.custom.cpp

```diff
@@ -1,2 +1,207 @@
+#include <server/converters.h>
+#include "custom/nixnet_converters.h"
+
 namespace nixnet_grpc {
-}  // namespace nixnet_grpc
\ No newline at end of file
+using nidevice_grpc::converters::convert_to_grpc;
+    
+::grpc::Status NiXnetService::ReadState(::grpc::ServerContext* context, const ReadStateRequest* request, ReadStateResponse* response)
+{
+    if (context->IsCancelled()) {
+        return ::grpc::Status::CANCELLED;
+    }
+    try {
+        auto session_ref_grpc_session = request->session_ref();
+        nxSessionRef_t session_ref = session_repository_->access_session(session_ref_grpc_session.id(), session_ref_grpc_session.name());
+        u32 state_id;
+        switch (request->state_id_enum_case()) {
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::kStateId: {
+                state_id = static_cast<u32>(request->state_id());
+                break;
+            }
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::kStateIdRaw: {
+                state_id = static_cast<u32>(request->state_id_raw());
+                break;
+            }
+            case nixnet_grpc::ReadStateRequest::StateIdEnumCase::STATE_ID_ENUM_NOT_SET: {
+                return ::grpc::Status(::grpc::INVALID_ARGUMENT, "The value for state_id was not specified or out of range");
+                break;
+            }
+        }
+
+        u32 state_size = request->state_size();
+        u64 time_stamp_100ns;
+        u32 comm_state;
+        FlexRayStats flex_ray_stats;
+        J1939CommState J1939_comm_state;
+        TimeLocalNetwork time_local_network;
+        void* state_value;
+
+        switch(state_id){
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_CURRENT : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_COMMUNICATING : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_START :{
+                state_value = &time_stamp_100ns;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_CAN_COMM :
+            case nixnet_grpc::ReadState::READ_STATE_STATE_FLEX_RAY_COMM :
+            case nixnet_grpc::ReadState::READ_STATE_STATE_LIN_COMM :
+            case nixnet_grpc::ReadState::READ_STATE_STATE_SESSION_INFO :{
+                state_value = &comm_state;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_FLEX_RAY_STATS :{
+                state_value = &flex_ray_stats;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_J1939_COMM :{
+                state_value = &J1939_comm_state;
+                break;
+            }
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_CURRENT_2 : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_COMMUNICATING_2 : 
+            case nixnet_grpc::ReadState::READ_STATE_STATE_TIME_START_2 :{
+                state_value = &time_local_network;
+                break;
+            }
```

Comment:
`default` should throw `INVALID_ARGUMENT` or something.

Reply:
Added default case